update: platform key differences page (#378)
JetBrains/kotlin-multiplatform-dev-docs: topics/compose/compose-platform-specifics.md

[//]: # (title: Default UI behavior on different platforms)

Compose Multiplatform aims to help you produce apps which behave as similarly as possible on different platforms.
On this page, you can learn about unavoidable differences or temporary compromises which
you should expect when writing shared UI code for different platforms
with Compose Multiplatform.

## Project structure

Regardless of the platform you're targeting, each of them needs a dedicated entry point:

> * For Android, that's the Activity whose job it is to show the main composable from common code.

Reviewer: ```suggestion
* For Android, that's the `Activity`, whose job is to show the main composable from common code.
```

* For an iOS app, that's the `@main` class or structure that initializes the app.
* For a JVM app, that's the `main()` function that starts the application which launches the main common composable.
* For a Kotlin/JS or Kotlin/Wasm app, that's the `main()` function that attaches the main common code composable
  to the web page.

Certain platform-specific APIs necessary for your app may not have multiplatform support,
and you will have to implement calling these APIs in platform-specific source sets.

> But before you do that, check out [klibs.io](https://klibs.io/), a JetBrains project which aims to comprehensively
> catalog all available Kotlin Multiplatform libraries.
> Already available are libraries for network code, databases, coroutines, and much more.

Reviewer: ```suggestion
Before doing so, check out [klibs.io](https://klibs.io/), a JetBrains project which aims to comprehensively
catalog all available Kotlin Multiplatform libraries.
There are already libraries available for network code, databases, coroutines, and much more.
```

## Input methods

### Software keyboards

> Each platform may handle software keyboards a little differently in the way it appears when a text field becomes active,
> for example.

Reviewer: ```suggestion
Each platform may handle software keyboards slightly differently, including the way a keyboard appears when a text field becomes active.
```

to account for [safe areas](https://developer.apple.com/documentation/UIKit/positioning-content-relative-to-the-safe-area).
Depending on your implementation, the software keyboard may be positioned a little differently on iOS.
Make sure to check that the keyboard does not cover important UI elements on both platforms. 

Compose Multiplatform currently doesn't support changing the default IME action, for example,
showing a magnifying glass or a checkmark instead of the usual &crarr; icon.

### Touch and mouse support

The current desktop implementation interprets all pointer manipulation as mouse gestures
and, therefore, does not support multitouch gestures.
For example, the common pinch-to-zoom gesture cannot be implemented with Compose Multiplatform for desktop
as it requires processing two touches at once.

## UI behavior and appearance

### Platform-specific functionality

Some common UI elements are not covered by Compose Multiplatform and cannot be customized using the framework.
Therefore, you should expect them to look different on different platforms, for example:

* **Share** actions that use system APIs to offer a list of specific ways to share or send content.

> * Native popup-views (for example, **Copy** or **Translate** actions for selected text).

Reviewer: ```suggestion
* Native pop-up views (for example, **Copy** or **Translate** actions for selected text).
```

### Scroll physics

For Android and iOS, the feel of scrolling is aligned with the platform.
For desktop, scrolling support is limited to the mouse wheel (as mentioned in [](#touch-and-mouse-support)). 

### Interop views

If you would like to embed native views within common composables, or vice versa,
you'll need to familiarize yourself with platform-specific mechanisms supported by Compose Multiplatform.

For iOS, there are separate guides for interop code with [SwiftUI](compose-swiftui-integration.md) and [UIKit](compose-uikit-integration.md).

For desktop, Compose Multiplatform supports [](compose-desktop-swing-interoperability.md).

### Back gesture

Android devices have back gesture support by default, and every screen reacts to the **Back** button in some fashion.

On iOS, there is no back gesture by default, although developers are encouraged to implement similar functionality
to meet user experience expectations.
Compose Multiplatform for iOS supports back gestures by default to mimic Android functionality.

> On desktop, Compose Multiplatform uses the **Esc** key by default as a back trigger.

Reviewer: ```suggestion
On desktop, Compose Multiplatform uses the **Esc** key as the default back trigger.
```

For details, see the [](compose-navigation.md#back-gesture) section.

### Text

With text, Compose Multiplatform doesn't guarantee pixel-perfect correspondence between different platforms:

* If you don't set a font explicitly, each system assigns a different default font to your text.
* Even for the same font, letter aliasing mechanisms specific to each platform can lead to noticeable differences.

> This doesn't have a significant impact on user experience (on the contrary, default fonts look as expected on each platform).

Reviewer: ```suggestion
This doesn't have a significant impact on user experience. On the contrary, the default fonts appear as expected on each platform.
```

> But pixel differences may get in the way of screenshot testing, for example.

Reviewer: ```suggestion
However, pixel differences may interfere with screenshot testing, for example.
```

### Initial performance

On iOS, you may notice a delay in the initial performance of individual screens compared to Android.
This can happen because Compose Multiplatform compiles UI shaders on demand.
So, if a particular shader is not cached yet, compiling it may delay rendering of a scene.

> This affects only the first time launches for each screen:
> after all necessary shaders are cached, their compilation won't delay subsequent launches.

Reviewer: ```suggestion
This issue affects only the first launch of each screen.
Once all necessary shaders are cached, subsequent launches are not delayed by compilation.
```

## Developer experience

### Previews

_Previews_ are non-interactable layout presentations of composables available in the IDE.

To see a preview for a composable:

1. Add an Android target to your project if there isn't one (the preview mechanism uses an Android library).
2. Mark composables that you would like to be previewable with the `@Preview` annotation in common code.
3. Switch to the **Split** or **Design** view in the editor window.
     It prompts you to build the project for the first time if you haven't done that yet.

In both IntelliJ IDEA and Android Studio, you will be able to see the initial layout of every composable
annotated with `@Preview` in the current file.

### Hot reload

> _Hot reload_ is the experience of the app reflecting changes in code without additional input, on the fly.

Reviewer: ```suggestion
_Hot reload_  refers to the app reflecting code changes on the fly without requiring additional input.
```

> For Compose Multiplatform, the hot reload functionality is available only for JVM (desktop) targets,
> but you can use it to quickly iron out problems before switching to your actual target platforms for fine-tuning.

Reviewer: ```suggestion
In Compose Multiplatform, hot reload functionality is available only for JVM (desktop) targets.
However, you can use it to quickly troubleshoot issues before switching to your intended platforms for fine-tuning.
```

for example.

Compose Multiplatform adopts the [Compose window insets approach](https://developer.android.com/develop/ui/compose/system/insets)
and imitates it on iOS
to account for [safe areas](https://developer.apple.com/documentation/UIKit/positioning-content-relative-to-the-safe-area).
Depending on your implementation, the software keyboard may be positioned a little differently on iOS.
Make sure to check that the keyboard does not cover important UI elements on both platforms. 

Compose Multiplatform currently doesn't support changing the default IME action, for example,
showing a magnifying glass or a checkmark instead of the usual &crarr; icon.

### Touch and mouse support

The current desktop implementation interprets all pointer manipulation as mouse gestures
and, therefore, does not support multitouch gestures.
For example, the common pinch-to-zoom gesture cannot be implemented with Compose Multiplatform for desktop
as it requires processing two touches at once.

## UI behavior and appearance

### System functionality

Some UI elements depend on system integration and may look different in the default configuration,
for example:

> * The "Share" actions use system APIs to offer a list of specific ways to share or send content.

Reviewer 2: Compose doesn't provide any API to request to `Share` something. I would remove it, as it out of the context of Compose

Author: That was my point, to highlight the fact that CMP doesn't cover everything one might call "UI": these are a couple of examples of things that are definitely going to be different. I rephrased the section to put this point forward, tell me if it's still off.

Reviewer 2: It is still off.

We describe functionality that doesn't exist unless users explicitly write it, not using Compose at all.

The native popup-views point on the other hand exists - one of the popups is triggered from the TextField.

Author: Okay, I think I misunderstood this in our conversation; I'm leaving pop-up views as a category example and text selection as a specific example.

* The popup-views (for example, for selected text) are fully native and currently cannot be customized
    through Compose Multiplatform.

### Scroll physics

For Android and iOS, the feel of scrolling is aligned with the platform.
For desktop, scrolling support is limited since the JVM target by default doesn't support

> common mobile gesture patterns like fling or overscroll (often used for the "pull to refresh" action).

Reviewer 3: "fling" is generally supported - snapping of pager is working based on it. Default scrolling shouldn't trigger it by design.
"pull to refresh" - is exact material component that uses explicitly touch move gestures. It's not about "platofrm support"

I'd better say that scrolling on Desktop is limited by mouse wheel and but it's mentioned in "Touch and mouse support" section

For desktop, Compose Multiplatform supports [](compose-desktop-swing-interoperability.md).

### Back gesture

Android devices have back gesture support by default, and every screen reacts to the **Back** button in some fashion.

On iOS, there is no back gesture by default, although developers are encouraged to implement similar functionality
to meet user experience expectations.
Compose Multiplatform for iOS supports back gestures by default to mimic Android functionality.

On desktop, Compose Multiplatform uses the **Esc** key by default as a back trigger.

For details, see the [](compose-navigation.md#back-gesture) section.

### Text

With text, Compose Multiplatform doesn't guarantee pixel-perfect correspondence between different platforms:

* If you don't set a font explicitly, each system assigns a different default font to your text.
* Even for the same font, letter aliasing mechanisms specific to each platform can lead to noticeable differences.

This doesn't have a significant impact on user experience (on the contrary, default fonts look as expected on each platform).
But pixel differences may get in the way of screenshot testing, for example.

> ### Initial performance

Reviewer 2: I would remove it if we don't have evidence and/or real cases. Benchmarks shows that [startup is okay](https://blog.jetbrains.com/kotlin/2025/05/compose-multiplatform-1-8-0-released-compose-multiplatform-for-ios-is-stable-and-production-ready/), which can be comparable to showing the screen.

Author: This is more about the principle than numbers — I didn't mean to imply that the initial performance is not okay, just that it works the way it works. I'll ask @pjBooms to weigh in here.

Reviewer 2: Regardless why this happens, it shouldn't a principle. Even if it happens, and we don't fix it, it is still an issue.

Reviewer 2: But if we want to describe this as an issue, we need evidence

Reviewer 2: > it is still an issue.

To understand the context - there are ways to fix this issue (Flutter fixed by replacing Skia, Skia has experimental backend, we also could provide some tooling for precompilation)

Author: I see your point, let's not talk about it as a baseline expectation for now. I'll clarify with Nikita, and maybe we'll include a section on this in the article on our benchmarks.
